crates/uv-requirements/src/pyproject.rs

```diff
@@ -1,133 +1,520 @@
+            // Include the optional dependencies if the extras are requested.
+            for (extra, optional_requirements) in &uv_requirements.optional_dependencies {
+                if extras.contains(extra) {
+                    used_extras.insert(extra.clone());
+                    requirements.extend(flatten_extra(
+                        &project.name,
+                        optional_requirements,
+                        &uv_requirements.optional_dependencies,
+                    ));
                 }
             }
         }
 
         Ok(Some(Self {
-            name,
+            name: project.name,
             requirements,
             used_extras,
         }))
     }
 }
 
+pub(crate) fn lower_requirements(
+    dependencies: &[String],
+    optional_dependencies: &IndexMap<ExtraName, Vec<String>>,
+    project_name: &PackageName,
+    project_dir: &Path,
+    project_sources: &HashMap<PackageName, Source>,
+    workspace_sources: &HashMap<PackageName, Source>,
+    workspace_packages: &HashMap<PackageName, String>,
+) -> Result<Requirements, Pep621Error> {
+    let dependencies = dependencies
+        .iter()
+        .map(|dependency| {
+            let requirement = pep508_rs::Requirement::from_str(dependency)?;
+            let name = requirement.name.clone();
+            lower_requirement(
+                requirement,
+                project_name,
+                project_dir,
+                project_sources,
+                workspace_sources,
+                workspace_packages,
+            )
+            .map_err(|err| Pep621Error::LoweringError(name, err))
+        })
+        .collect::<Result<_, Pep621Error>>()?;
+    let optional_dependencies = optional_dependencies
+        .iter()
+        .map(|(extra_name, dependencies)| {
+            let dependencies: Vec<_> = dependencies
+                .iter()
+                .map(|dependency| {
+                    let requirement = pep508_rs::Requirement::from_str(dependency)?;
+                    let name = requirement.name.clone();
+                    lower_requirement(
+                        requirement,
+                        project_name,
+                        project_dir,
+                        project_sources,
+                        workspace_sources,
+                        workspace_packages,
+                    )
+                    .map_err(|err| Pep621Error::LoweringError(name, err))
+                })
+                .collect::<Result<_, Pep621Error>>()?;
+            Ok((extra_name.clone(), dependencies))
+        })
+        .collect::<Result<_, Pep621Error>>()?;
+    Ok(Requirements {
+        dependencies,
+        optional_dependencies,
+    })
+}
+
+/// Combine `project.dependencies`/`project.optional-dependencies` with `tool.uv.sources`.
+pub(crate) fn lower_requirement(
+    requirement: pep508_rs::Requirement,
+    project_name: &PackageName,
+    project_dir: &Path,
+    project_sources: &HashMap<PackageName, Source>,
```

Comment:
In the sources design, what if we need to support multiple entries for a single package? Is it possible? What if I want to use a different index URL depending on whether I'm on mac or Linux?

Reply:
toml does not allow redefining keys, so we can't use the METADATA email method of repeating keys. In the future, we can allow making `Source` an array (through an untagged enum):
```toml
foo = [
    { url = "https://linux.example.org/foo-2.0.0-cp39-cp39-linux_x86_64.whl", os = "linux" },
    { url = "https://linux.example.org/foo-2.0.0-cp39-cp39-win_amd64.whl", os = "windows" }
]
```
I'd leave this as future work.